jenkins-job-components/src/main/java/com/ofg/job/component/JobComponent.java

```diff
@@ -0,0 +1,20 @@
+package com.ofg.job.component;
+
+import javaposse.jobdsl.dsl.Job;
+
+/**
+ * Interface for configurations applicable on Job.
+ */
+public interface JobComponent<J extends Job> {
+
+    JobComponent<? extends Job> EMPTY = job -> {};
+    
+    default <NJ extends J> JobComponent<NJ> and(JobComponent<NJ> other) {
```

Comment:
just an idea: I think a way to allow the user to decide on the order is possible. Another method that would accept `JobComponent<? super J>` but return without changing the `J` type. So `default JobComponent<J> nameMe(JobComponent<? super J> other) { ... }`. Given what operators are [overlodable in Groovy](http://groovy-lang.org/operators.html#Operator-Overloading) (and that `&&` doesn't seem to be overlodable) 
I'll use `>>` and `>>>` for the type-narrowing and type-retaining variants respectively in the example below:
 
```
AbstractJobConfigurer() >>
AbstractJobConfigurer() >>
MultiJobConfigurer() >>
MultiJobConfigurer() >>>
AbstractJobConfigurer() >>
MultiJobConfigurer()
```

WDYT?